docs/apis-tools/frontend-development/03-forms/03-customize-and-extend/03-integrate-api-data.md

```diff
@@ -0,0 +1,75 @@
+---
+id: integrate-api-data
+title: "Integrate API data"
+description: "Integrate external business data into your forms via APIs."
+---
+
+Read this page to learn how to integrate external business data into your forms via APIs.
+
+## Load data on form initation
+
+Before you initiate your form with data, make sure to fetch external business data and merge it with the process variables first. Data that is not bound to a form field using a key will not be submitted, keeping process instance data clean.
+
+As an example, use a `valuesExpression` in your form to populate the options of a select field.
+
+```js
+//...
+
+const schema = {
+  components: [
+    {
+      label: "Opportunities",
+      type: "select",
+      key: "opportunity",
+      valuesExpression: "=external.salesforce.opportunities"
+    }
+  ],
+  type: "default",
+  id: "TestForm",
+  schemaVersion: 12
+};
+
+const response = await fetch(url, fetchOptions);
+const opportunities = await response.json() //...
+
+// form context/input data
+const data = {
+  ...processVariables,
+  external: {
+    salesforce: {
+        opportunities
+    },
+    // ...
+  }
+};
+
+await form.importSchema(schema, data);
+```
+
+## Load data on runtime with form events
+
+:::info
+<span class="badge badge--platform">Workaround</span>
+
+Currently, there is no built-in way to update a form's context data on runtime. However, a workaround exists.
+:::
+
+To load and update data on runtime (e.g. when searching in a searchable select box, or entering a query in a text field), follow these steps:
+
+1. Listen to the `changed`, `formField.blur`, or `formField.search` event.
+2. Gather the current form state from the `changed` event, or call the `submit` function to retrieve the data.
+3. Find the query term in the changed state that is relevant for your API calls.
+4. Run your API call, e.g. fetch records based on the query term.
+5. Re-import the form schema but with the updated data (the current form state you obtained earlier, merged with the API results).
```

Comment:
I don't think you need to fully re-import, you can probably just form._setState().

Reply:
I don't think we should encourage using `_setState` since it's an internal API

Reply:
@Skaiir what does it take to expose setState to public, and is this generally a good idea? Let's discuss in our channel